### PR TITLE
Feature/optimistic sync

### LINKED ISSUE #395: Implement Optimistic Sync
### Describe the feature

We want to implement syncing. I think Optimistic Sync is how we should start https://github.com/ethereum/consensus-specs/blob/dev/sync/optimistic.md

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by alok-108**
### What was wrong?

Currently, the beacon node halts or rejects block imports if the execution layer is not yet synced or returns `SYNCING` / `ACCEPTED` (only `VALID` was permitted). Furthermore, historic block range syncing eagerly called the execution engine without optimistic synchronization support, slowing down sync and lacking a rollback mechanism for invalid execution payloads.

### How was it fixed?

- **Optimistic Sync Engine Handling**: Updated `verify_and_notify_new_payload` to return `PayloadStatusV1` and allow block imports on `VALID`, `ACCEPTED`, and `SYNCING`.
- **Optimistic Block Range Syncer**: Added `--optimistic-sync` support to `BlockRangeSyncer` with `is_optimistic_candidate_block` safe distance logic (128 slots) and `process_block_optimistic` for deferred EL verification.
- **Root Tracking**: Added `OptimisticRootsTable` to `BeaconDB` to track roots of optimistically imported blocks.
- **Invalid Payload Recovery & Rollback**: Implemented `handle_invalid_payload` to prune invalid blocks, dependent states, and column sidecars, back-tracking to `latest_valid_hash` without deadlock.
- **Testing**: Added unit and integration tests covering optimistic candidate filtering, invalid payload rollback, and table persistence.

Closes #395

### To-Do

- [x] I have read [CONTRIBUTING.md](https://github.com/ReamLabs/ream/blob/master/CONTRIBUTING.md).
- [x] This PR title follows [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/).

### CODE REVIEW

**Comment by KolbyML** [comment]
@perfogic is this PR fine for me to review, I seen the other 2 are blocked on your concern.

**Comment by perfogic** [comment]
@KolbyML Sorry for the late reply, I thought you were asking @alok-108.

Yes, you can. Our head sync currently only imports blocks when the EL returns `VALID`. 
With proper optimistic sync support, we could also import blocks with `SYNCING` or `ACCEPTED` responses.

**Comment by alok-108** [comment]
Thanks @perfogic and @KolbyML!

Exactly — with the changes in this PR:
1. `verify_and_notify_new_payload` now returns `PayloadStatusV1`, allowing blocks to be imported when the EL returns `SYNCING` or `ACCEPTED` rather than bailing on anything that isn't `VALID`.
2. Historic range sync supports `--optimistic-sync` via `process_block_optimistic` and tracks optimistic roots in `OptimisticRootsTable`.
3. An invalid payload cleanup/rollback mechanism (`handle_invalid_payload`) is in place if the EL later marks an imported block invalid.

Looking forward to your review @KolbyML, and happy to address any comments or adjustments!

**Comment by alok-108** [comment]
### Update: Implementation & Test Verification

Hey, quick update on this PR. I've pushed the remaining optimistic sync spec changes 

Here's what's new:

- **Fork choice now tracks optimistic roots on the fly.** We're passing the execution payload status (`Option<PayloadStatus>`) from `state_transition` into `on_block`. So when a block comes in with `SYNCING` / `ACCEPTED` (or when `skip_execution_validation = true`), it gets registered in `OptimisticRootsTable` right away.

- **Optimistic ancestors get resolved recursively.** Once the EL says a payload is `VALID`, we walk up the parent chain and unmark all its optimistic ancestors in `OptimisticRootsTable`.

- **The node syncing RPC is actually dynamic now.** `GET /eth/v1/node/syncing` checks `db.optimistic_roots_provider().get(head)` instead of just returning `false` all the time.

- **Fixed the gossipsub validation.** The payload status pattern matching in `beacon_block.rs` was off, so I cleaned that up.

---

### Tests

Ran everything locally, and it's all passing:

- **`ream-chain-beacon`**  
  `cargo test --features devnet5 -p ream-chain-beacon`  
  **5/5 passed:**
  - `test_optimistic_root_storage`
  - `test_optimistic_ancestor_resolution` (checks that ancestors get resolved on a valid payload)
  - `test_handle_invalid_payload_cleanup`
  - `test_invalid_payload_multi_descendant_rollback`
  - `test_handle_invalid_payload_unimported_block_preserves_head`

- **`ream-syncer`**  
  `cargo test --features devnet5 -p ream-syncer`  
  **6/6 passed:**
  - Distance check with `SAFE_SLOTS_TO_IMPORT_OPTIMISTICALLY`
  - Range and block cache fetch tests

- **Checks:**
  - `cargo check --features devnet5 -p ream-fork-choice-beacon` — clean
  - `cargo check --features devnet5 -p ream-rpc-beacon` — clean
  - `cargo check --features devnet5 -p ream-network-manager` — clean

Branch is rebased, tests are green, and this PR is **ready for review and merge** 🚀

**Comment by alok-108** [comment]
Hey @KolbyML 👋
Hope you're doing well! Just a gentle check-in on this PR — totally understand you're busy. Any chance you might be able to take a look sometime? No rush at all, just wanted to get a rough idea. Everything's passing locally and it's ready from my side.
Thanks a lot!

File: crates/common/chain/beacon/src/beacon_chain.rs [hunk 1]
```diff
@@ -53,72 +60,83 @@ impl BeaconChain {
...
+                        .unwrap_or(B256::ZERO)
+                } else {
+                    B256::ZERO
+                };
+                self.handle_invalid_payload(store, block_root, latest_valid)
+                    .await?;
+                bail!("Block payload is invalid");
+            }
+            return Err(e);
+        }
```
Comment (KolbyML) [inline crates/common/chain/beacon/src/beacon_chain.rs]: Err(err) and we shouldn't be parsing errors like this, we should be doing type checks

File: crates/common/chain/beacon/src/beacon_chain.rs [hunk 2]
```diff
@@ -53,72 +60,83 @@ impl BeaconChain {
...
+        on_block(
+            &mut store,
+            &signed_block,
+            &self.execution_engine,
+            signed_block.message.slot >= beacon_network_spec().slot_n_days_ago(17),
+            true, // skip execution validation
+        )
+        .await?;
+
+        // Insert the root as optimistic in the database
```
Comment (KolbyML) [inline crates/common/chain/beacon/src/beacon_chain.rs]: these commands don't add anything, they should be removed

File: crates/common/chain/beacon/src/beacon_chain.rs [hunk 3]
```diff
@@ -139,6 +157,180 @@ impl BeaconChain {
...
+            // Remove from all DB tables so filter_block_tree() stays consistent
+            store.db.block_provider().remove(root)?;
+            store.db.state_provider().remove(root)?;
+            let _ = store.db.slot_index_provider().remove(slot);
+            let _ = store
+                .db
+                .parent_root_index_multimap_provider()
+                .remove_child(parent_root, root);
+            let _ = store.db.unrealized_justifications_provider().remove(root);
+            let _ = store.db.optimistic_roots_provider().remove(root);
```
Comment (KolbyML) [inline crates/common/chain/beacon/src/beacon_chain.rs]: why aren't we handling results, please fix this across the whole pr

File: crates/common/chain/beacon/src/beacon_chain.rs [hunk 4]
```diff
@@ -139,6 +157,180 @@ impl BeaconChain {
...
+
+        let justified_checkpoint = store.db.justified_checkpoint_provider().get()?;
+        let safe_block_hash = if justified_checkpoint.root == B256::ZERO {
+            B256::ZERO
+        } else {
+            store
+                .db
+                .block_provider()
+                .get(justified_checkpoint.root)?
+                .map(|b| b.message.body.execution_payload.block_hash)
```
Comment (KolbyML) [inline crates/common/chain/beacon/src/beacon_chain.rs]: what is b

**Review by KolbyML — COMMENTED** [review]
I left 4 comments, the same issues reoccur a lot in the pR, please fix them all then rerequest my review again

**Comment by alok-108** [comment]
Thanks @KolbyML for the detailed review. I've addressed all four points in `9fd9a41`:

1. **Typed errors**: Added an `EngineError` enum with `InvalidPayload { latest_valid_hash: Option<B256> }` and `InvalidBlockHash`. `process_execution_payload` now returns these instead of string bail, and `process_block` uses `err.downcast_ref::<EngineError>()` for type-safe matching.

2. **Dead code**: Removed the redundant optimistic roots insertion and the stale comment from `process_block_optimistic`. `on_block` already handles it when `skip_execution_validation = true`.

3. **DB results**: Replaced all `let _ =` and `.ok()` discards with `?` propagation across `beacon_chain.rs`, `handlers.rs`, `store.rs`, and `syncing.rs`. File deletions handle `NotFound` idempotently while bubbling real I/O errors.

4. **Naming**: Renamed `b` to `beacon_block`, `e` to `err`, and fixed the `execution_enigne` typo to `execution_engine`.

Also added tests for the `EngineError` downcast and the `latest_valid_hash: None` rollback. All tests green locally:

- `ream-chain-beacon`: 5/5 pass
- `ream-syncer`: 6/6 pass
- `cargo clippy` and `cargo fmt --check`: clean

Thanks again for the review.
